kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,1242 @@
+                    self.check_padding_requirements();
+                    match self.padding_requirement.get() {
+                        // if we decided we need to write a padding app after the new app, we go ahead and do it
+                        Some(PaddingRequirement::PostPad)
+                        | Some(PaddingRequirement::PreAndPostPad) => {
+                            // calculating the distance between our app and either the next app
+                            let new_app_end_address = new_app_start_address + app_length;
+                            let post_pad_length =
+                                self.next_app_start_addr.get() - new_app_end_address;
+
+                            let padding_result =
+                                self.write_padding_app(post_pad_length, new_app_end_address);
+                            let _ = match padding_result {
+                                Ok(()) => Ok(()),
+                                Err(e) => {
+                                    self.reset_process_loading_metadata();
+                                    Err(e) // this means we were unable to write the padding app
+                                }
+                            };
+                        }
+                        // Otherwise we let the client know we are done with the setup, and we are ready to write the app to flash
+                        None | Some(PaddingRequirement::PrePad) => {
+                            self.state.set(State::FirstWrite);
+                            self.client.map(|client| {
+                                client.setup_done();
+                            });
+                        }
+                    };
+                    Ok(app_length)
+                }
+                Err(_err) => {
+                    self.reset_process_loading_metadata(); // reset the state to None because we did not find any available address for this app
+                    Err(ErrorCode::FAIL)
+                }
+            }
+        } else {
+            Err(ErrorCode::BUSY) // this means the kernel is busy doing some other operation already.
+        }
+    }
+
+    fn write_app_data(
+        &self,
+        buffer: SubSliceMut<'static, u8>,
+        offset: usize,
+        _processid: ProcessId,
+    ) -> Result<(), ErrorCode> {
+        match self.state.get() {
+            Some(State::FirstWrite) => {
+                if offset != self.next_offset.get() {
+                    // self.next_offset should be 0 at this stage.
+                    self.reset_process_loading_metadata(); // clear out new app size, start address, current offset and state
+                    return Err(ErrorCode::INVAL); // this means the app is trying to write to an address that is not the
+                                                  // start address assigned to it after setup. Return an Invalid error.
+                }
```

Comment:
Oh I see. I still think it's simpler to just go based on offset rather than tracking first write. And there is still the issue of writing offset 0 later in the process.